### ISSUE TITLE
[BUG] cython dataclass does not generate signature for the class

### ISSUE THREAD

**Issue by stestagg** [cython/cython#5722]
### Describe the bug

A python dataclass can be passed to `inspect.signature(Cls)` and a signature (I assume of the generated __init__ method) is returned.

For a cython cdef dataclass, the `inspect.signature(CyCls)` call raises an exception.

### Code to reproduce the behaviour:

testdc.pyx:
```cython
import dataclasses
from cython.dataclasses cimport dataclass

@dataclass(frozen=True)
cdef class CyDC:
    test: int = 0

@dataclasses.dataclass(frozen=True)
class PyDC:
    test: int = 0

```

test.py:
```
import inspect
import pyximport; pyximport.install()

if __name__ == '__main__':
    import testdc
    print(">>>> PYTHON dataclass <<<<<")
    print(inspect.signature(testdc.PyDC))
    print(">>>> CYTHON dataclass <<<<<")
    print(inspect.signature(testdc.CyDC))
```

### Expected behaviour

Expected:
```
>>>> PYTHON dataclass <<<<<
(test: 'int' = 0) -> None
>>>> CYTHON dataclass <<<<<
(test: 'int' = 0) -> None
```

Actual:
```
>>>> PYTHON dataclass <<<<<
(test: 'int' = 0) -> None
>>>> CYTHON dataclass <<<<<
Traceback (most recent call last):
...
ValueError: no signature found for builtin type <class 'testdc.CyDC'>
```

### OS

macOs

### Python version

3.10.4

### Cython version

3.0.2

**Comment by da-woods** [comment]
I suspect this applies to any `cdef class` - if we were to do it I doubt that it should be specific to dataclasses.

Obviously there's the usual caveat that `cdef classes` trade some convenience for efficiency so not everything that works for a regular class is necessarily expected to work for them.

**Comment by stestagg** [comment]
You're right (about it applying to any cdef class), I just adapted/checked the above. 

Interestingly, in my case, the reason I hit this is because `pdoc` was failing horribly just for dataclasses, not for other cython cclasses.  This turns out to be because it has a special branch in the code for dataclasses (not because of cython treating dataclasses differently)

Would be a big nice-to-have, but I can see that I hit a corner-case here.

**Comment by da-woods** [comment]
That suggests we've been a little too successful at disguising Cython dataclasses as dataclasses!

It's possibly a little fiddly because we have two methods that can do class initialization (`__cinit__` and `__init__`) so need to get information from two places. But I doubt it's impossible to generate a signature.

**Comment by seberg** [comment]
Just for reference, and maybe someone is interested in it.  @jorenham used the text-signature path to work around this in NumPy: https://github.com/numpy/numpy/pull/30164 which seems at least an OK work-around for those caring about a signature.
But adding it automatically, would be nicer of course (and maybe text signature would even be fine via injecting it automatically into the docstring).
